src/Config/ConfigLocator.php

```diff
@@ -400,24 +401,76 @@ public function getSiteAliasPaths($paths, Environment $environment)
      *
      * @return array
      */
-    public function getCommandFilePaths($commandPaths)
+    public function getCommandFilePaths($commandPaths, $root)
     {
-        // Start with the built-in commands.
-        $searchpath = [
+        $builtin = $this->getBuiltinCommandFilePaths();
+        $included = $this->getIncludedCommandFilePaths($commandPaths);
+        $site = $this->getSiteCommandFilePaths(["$root/drush", dirname($root) . '/drush']);
+
+        return array_merge(
+            $builtin,
+            $included,
+            $site
+        );
+    }
+
+    /**
+     * Return all of the built-in commandfile locations
+     */
+    protected function getBuiltinCommandFilePaths()
+    {
+        return [
             dirname(__DIR__),
         ];
+    }
 
+    /**
+     * Return all of the commandfile locations specified via
+     * an 'include' option.
+     */
+    protected function getIncludedCommandFilePaths($commandPaths)
+    {
+        $searchpath = [];
         // Commands specified by 'include' option
         foreach ($commandPaths as $commandPath) {
             if (is_dir($commandPath)) {
                 $searchpath[] = $commandPath;
             }
         }
-
         return $searchpath;
     }
 
     /**
+     * Return all of the commandfile paths in any '$root/drush' or
+     * 'dirname($root)/drush' directory that contains a composer.json
+     * file or a 'Commands' or 'src/Commands' directory.
+     */
+    protected function getSiteCommandFilePaths($directories)
+    {
+        $result = [];
+
+        $directories = array_filter($directories, 'is_dir');
+
+        if (empty($directories)) {
+            return $result;
+        }
+
+        // Find projects
+        $finder = new Finder();
+        $finder->files()
+            ->ignoreUnreadableDirs()
+            ->path('#composer.json$|^src/Commands|^Commands#')
+            ->in($directories)
+            ->depth('<= 3');
```

Comment:
is there a chance we scan large dirs like node_modules and such?

Reply:
This only scans:
- The contents of the ./drush folder
- The contents of the ./drush/contrib folder (or whatever it is called)
- The contents of each ./drush/contrib folder (all the individual extensions)

If someone manually commits an extension into ./drush, then it is possible that a drush extension might be found immediately inside ./drush.  It is expected that only policy files should be installed like this; larger sitewide extensions (that are not modules) should be installed via composer.

If someone adds a directory with 20,000 files immediately inside ./drush, then we will scan it. Don't do that. Only Drush things should be inside ./drush, so it shouldn't be a problem.

We could, perhaps, exclude `site-aliases` and other special Drush folder names.